Require syntactically-valid blockchains in functional and unit tests

TheBlueMatt opened the pull request:
This is an expansion of #831 and #838 will be rebased on it soon.

jkczyz commented on lightning/src/ln/functional_test_utils.rs:
```diff
@@ -85,12 +94,28 @@ pub fn connect_block<'a, 'b, 'c, 'd>(node: &'a Node<'b, 'c, 'd>, block: &Block,
+pub fn disconnect_blocks<'a, 'b, 'c, 'd>(node: &'a Node<'b, 'c, 'd>, count: u32) {
+	assert!(node.blocks.lock().unwrap().len() as u32 > count); // Cannot disconnect genesis
+	for _ in 0..count {
+		let (block_header, height) = {
+			let blocks = node.blocks.lock().unwrap();
+			(blocks[blocks.len() - 1].0, blocks[blocks.len() - 1].1)
+		};
```
Wondering if this could be simplified a bit without the mutex. Is there a need for it?

TheBlueMatt replied:
Oops, indeed. We just need interior mutability, so I swapped it for a refcell.

jkczyz commented on lightning/src/ln/functional_test_utils.rs:
```diff
@@ -85,12 +85,13 @@ pub fn connect_block<'a, 'b, 'c, 'd>(node: &'a Node<'b, 'c, 'd>, block: &Block,
 pub fn disconnect_block<'a, 'b, 'c, 'd>(node: &'a Node<'b, 'c, 'd>, header: &BlockHeader, height: u32) {
 	node.chain_monitor.chain_monitor.block_disconnected(header, height);
 	node.node.block_disconnected(header);
-	node.node.test_process_background_events();
```
Was removing this intentional?

TheBlueMatt replied:
Yes, though it happened int he wrong commit, I moved it to a later commit and since no previous tests relied on this behavior, removing this for `test_unconf_chan` to use `disconnect_block` is nice.

jkczyz replied:
Looks like the commit message for d1d1d1fb76fe4b3afb642392af33a09d3ba34dc3 doesn't reference a valid commit?

TheBlueMatt replied:
Oh oops, its the immediately-preceeding "f" commit - it removes this hunk after "Add assertions for in-order block [dis]connection in ChannelManager"

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -520,7 +520,7 @@ pub(crate) const MAX_LOCAL_BREAKDOWN_TIMEOUT: u16 = 2 * 6 * 24 * 7;
 /// the HTLC via a full update_fail_htlc/commitment_signed dance before we hit the
 /// CLTV_CLAIM_BUFFER point (we static assert that it's at least 3 blocks more).
-const CLTV_EXPIRY_DELTA: u16 = 6 * 12; //TODO?
+const CLTV_EXPIRY_DELTA: u16 = 6 * 6; //TODO?
```
Won't this affect more than just the tests? If this was intentional, could you make a separate commit explaining the change?

TheBlueMatt replied:
See #849.

jkczyz commented on lightning/src/ln/functional_test_utils.rs:
```diff
@@ -853,13 +893,13 @@ macro_rules! expect_payment_failed {
-				assert!(error_data.is_some());
+				assert_eq!($expected_payment_hash, *payment_hash, "unexpected payment_hash");
+				assert_eq!($rejected_by_dest, rejected_by_dest, "unexpected rejected_by_dest value");
+				assert!(error_code.is_some(), "expected error_code.is_some() = true");
+				assert!(error_data.is_some(), "expected error_data.is_some() = true");
 				$(
-					assert_eq!(error_code.unwrap(), $expected_error_code);
-					assert_eq!(&error_data.as_ref().unwrap()[..], $expected_error_data);
+					assert_eq!($expected_error_code, error_code.unwrap(), "unexpected error code");
+					assert_eq!($expected_error_data, &error_data.as_ref().unwrap()[..], "unexpected error data");
 				)*
```
Any reason to change the ordering? I tend to think leaving the expected last is more readable in code. Not sure if the error message is suppose to correspond with either side despite being printed on the same line as the second.

valentinewallace replied:
Interesting, I find `expected: {}, actual: {}` more readable (I think I made this change originally).

jkczyz replied:
FWIW, my reading is "assert that X has expected value Y".

jkczyz commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -2526,8 +2496,7 @@ fn test_justice_tx() {
 	{
-		let mut header = BlockHeader { version: 0x20000000, prev_blockhash: Default::default(), merkle_root: Default::default(), time: 42, bits: 42, nonce: 42 };
-		connect_block(&nodes[1], &Block { header, txdata: vec![revoked_local_txn[0].clone()] }, 1);
+		confirm_transaction_at(&nodes[1], &revoked_local_txn[0], nodes[1].best_block_info().1 + 1);
```
Would it be possible to add another helper to use when needing a height that is offset from the best block to avoid explicitly calling `best_block_info()`?

TheBlueMatt commented:
Fixed a few comments and cleaned up the util method interface more, sadly it required reworking the commits so didn't leave fixup commits in between. Will respond to the further comments tomorrow.

valentinewallace commented on lightning/src/ln/functional_test_utils.rs:
```diff
@@ -44,31 +44,50 @@ use std::sync::Mutex;
 use std::collections::HashMap;
 
-pub const CHAN_CONFIRM_DEPTH: u32 = 100;
+pub const CHAN_CONFIRM_DEPTH: u32 = 10;
```
This makes test logs much nicer to read :+1:

TheBlueMatt commented:
Rebased on main/#849.

jkczyz commented on lightning/src/ln/functional_tests.rs:
```diff
@@ -2965,11 +2965,7 @@ fn test_htlc_on_chain_timeout() {
-	cfg.channel_options.cltv_expiry_delta = 12*6;
-	cfg.channel_options.announced_channel = true;
-	let node_chanmgrs = create_node_chanmgrs(3, &node_cfgs, &[Some(cfg.clone()), Some(cfg.clone()), Some(cfg.clone())]);
+	let node_chanmgrs = create_node_chanmgrs(3, &node_cfgs, &[None, None, None]);
```
Unclear what 444c154d6e660ad4228d0ea7b04ce72d7a2c1c01 is a fixup for.

TheBlueMatt replied:
It reverts a hunk from #849 - I could probably not revert it, but I'd need to redo the test changes here to figure out how it used to be.

jkczyz commented on lightning/src/ln/functional_test_utils.rs:
```diff
@@ -82,17 +82,12 @@ pub fn confirm_transaction_at<'a, 'b, 'c, 'd>(node: &'a Node<'b, 'c, 'd>, tx: &T
-	if parent { assert_eq!(prev_blockhash, node.best_block_hash()); }
-	assert_eq!(node.best_block_info().1, height);
-
+pub fn connect_blocks<'a, 'b, 'c, 'd>(node: &'a Node<'b, 'c, 'd>, depth: u32) -> BlockHash {
```
Could you also remove `height` from `connect_block` and `disconnect_block`? These are also called in the tests.

jkczyz reviewed (commented):
Liking how `functional_test_utils` is shaping up! Definitely makes the tests easier to read. Left one related comment.

I couldn't quite follow what some of the fixups corresponded to assuming they were going to be squashed with their parent commits. Left a couple comments.

TheBlueMatt commented:
Rebased after merge of 848.

jkczyz commented on lightning/src/ln/functional_test_utils.rs:
```diff
@@ -79,49 +79,44 @@ pub fn confirm_transaction_at<'a, 'b, 'c, 'd>(node: &'a Node<'b, 'c, 'd>, tx: &T
+	node.chain_monitor.chain_monitor.block_disconnected(header, node.best_block_info().1);
 	node.node.block_disconnected(header);
 	node.blocks.borrow_mut().pop();
 }
```
Is passing the header redundant since it should be the popped block header?

TheBlueMatt replied:
Indeed, it does not, luckily it is also trivial to drop it :).

jkczyz reviewed (approved):
LGTM aside from one minor comment.

TheBlueMatt commented:
Note this depends on #849 which needs to go first.

TheBlueMatt commented:
Squashed. The diff from val's last ACK rebased onto main (`3613425a` available on my `2021-03-80fcdd3-rebased` branch) is the function argument drops that Jeff suggested. The only diff from the last review by jeff is one function drop and
```
$ git diff-tree -U1 fb1ecd6  a61289af0b717c70b00ac880cd155933420596b2 lightning/src/ln/functional_test_utils.rs
diff --git a/lightning/src/ln/functional_test_utils.rs b/lightning/src/ln/functional_test_utils.rs
index 2485f3d3..e7f61c6a 100644
--- a/lightning/src/ln/functional_test_utils.rs
+++ b/lightning/src/ln/functional_test_utils.rs
@@ -1203,2 +1203,5 @@ pub fn create_node_chanmgrs<'a, 'b>(node_count: usize, cfgs: &'a Vec<NodeCfg<'b>
 		let mut default_config = UserConfig::default();
+		// Set cltv_expiry_delta slightly lower to keep the final CLTV values inside one byte in our
+		// tests so that our script-length checks don't fail (see ACCEPTED_HTLC_SCRIPT_WEIGHT).
+		default_config.channel_options.cltv_expiry_delta = 6*6;
 		default_config.channel_options.announced_channel = true;
```
